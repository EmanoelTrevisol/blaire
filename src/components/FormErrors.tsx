import React from 'react';
import { StyleSheet, View, Text } from 'react-native';

export interface IFormErrorProps {
  messages: {
    required: string;
    valid: string;
  };
  error: boolean;
  errors: {
    valid: boolean;
    required: boolean;
  };
  dirty: boolean;
}

const FormErrors = (props: IFormErrorProps) => {
  const {
    error,
    dirty,
    messages,
    errors: { required, valid },
  } = props;
  if (!error || !dirty) {
    return null;
  }

  return (
    <View style={stl.errors}>
      {(required && (
        <Text style={stl.inputErrorText} testID="required-error">
          {messages.required}
        </Text>
      )) ||
        null}
      {(valid && (
        <Text style={stl.inputErrorText} testID="valid-error">
          {messages.valid}
        </Text>
      )) ||
        null}
    </View>
  );
};

const stl = StyleSheet.create({
  errors: {
    marginVertical: 10,
  },
  inputErrorText: {
    color: '#dd0000',
    fontSize: 12,
  },
});

export default FormErrors;
