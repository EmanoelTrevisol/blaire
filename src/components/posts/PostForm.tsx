import React, { useRef, useReducer, useEffect } from 'react';
import { StyleSheet, View, Text, TextInput } from 'react-native';
import { cloneDeep } from 'lodash';

import { FormActionTypes, ActionTypes } from '@store/form/types';
import { formReducer } from '@store/form/reducer';
import initialFormState from '@store/form/state';
import LoaderButton from '../LoaderButton';
import FormErrors from '@components/FormErrors';
import { Post } from '@models/Post';

const modifiedInitialFormState = {
  ...initialFormState,
  errors: {
    ...initialFormState.errors,
    minLength: true,
    maxLength: false,
  },
};

const titleErrorMessages = {
  required: 'Ops... tá vazio aqui.',
  minLength: 'Ops... conta um pouco mais pra gente.',
  maxLength:
    'Ops... assim tá muito comprido. Precisamos que tenha no máximo 35 caracteres.',
};

const bodyErrorMessages = {
  required: 'Ops... tá vazio aqui.',
  minLength: 'Ops... conta um pouco mais pra gente.',
  maxLength:
    'Ops... assim tá muito comprido. Precisamos que tenha no máximo 280 caracteres.',
};

const titleLengthValidations = {
  min: 5,
  max: 35,
};

const bodyLengthValidations = {
  min: 30,
  max: 280,
};

function validateField({
  dispatcher,
  value,
  lengthLimits,
}: {
  dispatcher: React.Dispatch<FormActionTypes>;
  value: string;
  lengthLimits: typeof titleLengthValidations | typeof bodyLengthValidations;
}) {
  dispatcher({
    type: ActionTypes.SET_DIRTY,
    payload: {
      dirty: true,
    },
  });

  let error = false;
  const errors = {
    required: false,
    minLength: false,
    maxLength: false,
  };

  if (!value) {
    error = true;
    errors.required = true;
    errors.maxLength = false;
    errors.minLength = false;
  } else if (value.length < lengthLimits.min) {
    error = true;
    errors.required = false;
    errors.minLength = true;
    errors.maxLength = false;
  } else if (value.length > lengthLimits.max) {
    error = true;
    errors.required = false;
    errors.minLength = false;
    errors.maxLength = true;
  } else {
    error = false;
    errors.required = false;
    errors.maxLength = false;
    errors.minLength = false;
  }

  dispatcher({
    type: ActionTypes.SET_ERRORS,
    payload: {
      error,
      errors,
    },
  });
}

export interface IAuthFormProps {
  isCreate?: boolean;
  onSubmit: Function;
  buttonTitle: string;
  isSubmitting: boolean;
  post: Post | null;
}

const PostForm = (props: IAuthFormProps) => {
  const [titleState, titleDispatcher] = useReducer(
    formReducer,
    cloneDeep(modifiedInitialFormState),
  );
  const [bodyState, bodyDispatcher] = useReducer(
    formReducer,
    cloneDeep(modifiedInitialFormState),
  );

  const titleInputRef = useRef();

  const { onSubmit, buttonTitle, isSubmitting, isCreate } = props;

  useEffect(() => {
    if (isCreate) {
      titleInputRef.current?.focus();
    }
  }, [isCreate]);

  const submit = async () => {
    validateTitle();
    validateBody();
    if (titleState.error || bodyState.error) {
      return;
    }

    try {
      await onSubmit({
        title: titleState.value,
        body: bodyState.value,
      });
    } catch (error) {
      console.log('Error', error);
    }
  };

  const validateTitle = () => {
    validateField({
      dispatcher: titleDispatcher,
      value: titleState.value,
      lengthLimits: titleLengthValidations,
    });
  };

  const validateBody = () => {
    validateField({
      dispatcher: bodyDispatcher,
      value: bodyState.value,
      lengthLimits: bodyLengthValidations,
    });
  };

  return (
    <>
      <View style={stl.form}>
        <View style={stl.formControl}>
          <View style={stl.label}>
            <Text style={stl.labelText}>Título</Text>
          </View>
          <TextInput
            ref={titleInputRef}
            autoCapitalize="none"
            style={stl.input}
            value={titleState.value}
            editable={!isSubmitting}
            keyboardType="email-address"
            textContentType="username"
            placeholder="Crie seu título"
            placeholderTextColor="#898989"
            onBlur={validateTitle}
            onChangeText={(value) =>
              titleDispatcher({
                type: ActionTypes.SET_VALUE,
                payload: { value },
              })
            }
          />
          <FormErrors messages={titleErrorMessages} {...titleState} />
        </View>
        <View style={stl.formControl}>
          <View style={stl.label}>
            <Text style={stl.labelText}>Post</Text>
          </View>
          <TextInput
            multiline={true}
            numberOfLines={5}
            style={stl.input}
            value={bodyState.value}
            editable={!isSubmitting}
            placeholder="Mal posso esperar para ler o que você tem para contar"
            placeholderTextColor="#898989"
            onChangeText={(value) =>
              bodyDispatcher({
                type: ActionTypes.SET_VALUE,
                payload: { value },
              })
            }
            onBlur={validateBody}
          />
          <FormErrors messages={bodyErrorMessages} {...bodyState} />
        </View>
      </View>
      <View style={stl.actions}>
        <LoaderButton
          buttonStyle={stl.button}
          isSubmitting={isSubmitting}
          onSubmit={submit}
        >
          <Text style={stl.buttonText}>{buttonTitle}</Text>
        </LoaderButton>
      </View>
    </>
  );
};

const stl = StyleSheet.create({
  form: {
    marginBottom: 60,
  },
  formControl: {
    maxHeight: 80,
    marginVertical: 20,
  },
  label: {
    marginVertical: 5,
  },
  labelText: {
    fontSize: 20,
  },
  input: {
    paddingVertical: 5,
    borderBottomColor: '#000',
    borderBottomWidth: 1,
  },
  actions: {},
  button: {
    backgroundColor: '#ea94a5',
    padding: 16,
    borderRadius: 50,
    height: 50,
  },
  buttonText: {
    textAlign: 'center',
    fontWeight: 'bold',
  },
  labelInstructions: {
    fontSize: 12,
    color: '#232323',
  },
});

export default PostForm;
