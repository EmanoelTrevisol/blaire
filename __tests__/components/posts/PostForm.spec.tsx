import React from 'react';
import PostForm, { IProps } from '@components/posts/PostForm';
import FormErrors from '@/components/form/FormErrors';
import { act } from 'react-test-renderer';
import { fireEvent, render, waitFor } from '@testing-library/react-native';

const TITLE_INPUT_PLACEHOLDER = 'Crie seu título';
const BODY_INPUT_PLACEHOLDER =
  'Mal posso esperar para ler o que você tem para contar';

let props: IProps;

describe('PostForm.tsx', () => {
  const renderComp = (cpProps: IProps) => {
    return render(<PostForm {...cpProps} />);
  };

  beforeEach(() => {
    props = {
      buttonTitle: 'Enviar',
      onSubmit: jest.fn(),
    };
  });

  afterEach(() => {
    jest.resetAllMocks();
  });

  test('matches snapshot', () => {
    const { toJSON } = renderComp(props);

    expect(toJSON()).toMatchSnapshot();
  });

  describe('inputs and submition behavior', () => {
    describe('does not call onSubmit', () => {
      test('when title and body empty', async () => {
        const { getByText } = renderComp(props);

        const submitButton = getByText(props.buttonTitle);
        await waitFor(() => fireEvent.press(submitButton));

        expect(props.onSubmit).not.toHaveBeenCalled();
      });

      test('when only title is empty', async () => {
        const { queryByPlaceholderText, getByText } = renderComp(props);

        const submitButton = getByText(props.buttonTitle);
        const bodyInput = queryByPlaceholderText(BODY_INPUT_PLACEHOLDER)!;

        act(() => fireEvent.changeText(bodyInput, 'EssaEH4n0v4s3nh4'));
        act(() => fireEvent(bodyInput, 'blur'));

        await waitFor(() => fireEvent.press(submitButton));

        expect(props.onSubmit).not.toHaveBeenCalled();
      });

      test('when body is invalid', async () => {
        const { queryByPlaceholderText, getByText } = renderComp(props);

        const submitButton = getByText(props.buttonTitle);
        const bodyInput = queryByPlaceholderText(BODY_INPUT_PLACEHOLDER)!;

        act(() => fireEvent.changeText(bodyInput, 'abc'));
        act(() => fireEvent(bodyInput, 'blur'));

        await waitFor(() => fireEvent.press(submitButton));

        expect(props.onSubmit).not.toHaveBeenCalled();
      });

      test('when title is invalid', async () => {
        const { queryByPlaceholderText, getByText } = renderComp(props);

        const submitButton = getByText(props.buttonTitle);
        const titleInput = queryByPlaceholderText(TITLE_INPUT_PLACEHOLDER)!;

        act(() => fireEvent.changeText(titleInput, 'title'));
        act(() => fireEvent(titleInput, 'blur'));

        await waitFor(() => fireEvent.press(submitButton));

        expect(props.onSubmit).not.toHaveBeenCalled();
      });
    });

    describe('calls onSubmit', () => {
      test('with valid title and body', async () => {
        props.onSubmit = jest.fn(() => Promise.resolve());

        const { queryByPlaceholderText, getByText } = renderComp(props);

        const args = {
          title: 'Esse é um título de POST',
          body: 'Esse post é muito informativo e é bem interessante',
        };

        // const submitButton = ;
        const titleInput = queryByPlaceholderText(TITLE_INPUT_PLACEHOLDER)!;
        const bodyInput = queryByPlaceholderText(BODY_INPUT_PLACEHOLDER)!;

        act(() => fireEvent.changeText(titleInput, args.title));
        act(() => fireEvent(titleInput, 'blur'));
        act(() => fireEvent.changeText(bodyInput, args.body));
        act(() => fireEvent(bodyInput, 'blur'));

        await act(() => fireEvent.press(getByText(props.buttonTitle)));

        expect(props.onSubmit).toHaveBeenCalledWith(args);
      });
    });

    describe('sends correct error props to FormErrors', () => {
      describe('title errors', () => {
        test('required', () => {
          const { getByPlaceholderText, UNSAFE_getAllByType } = renderComp(
            props,
          );

          const titleInput = getByPlaceholderText(TITLE_INPUT_PLACEHOLDER);
          act(() => fireEvent(titleInput, 'blur'));

          const formErrors = UNSAFE_getAllByType(FormErrors);

          const formErrorProps = formErrors[0].props;

          expect(formErrorProps.dirty).toBe(true);
          expect(formErrorProps.error).toBe(true);
          expect(formErrorProps.errors.required).toBe(true);
          expect(formErrorProps.errors.minLength).toBe(false);
          expect(formErrorProps.errors.maxLength).toBe(false);
        });
        test('minLength', () => {
          const { getByPlaceholderText, UNSAFE_getAllByType } = renderComp(
            props,
          );

          const titleInput = getByPlaceholderText(TITLE_INPUT_PLACEHOLDER);

          act(() => fireEvent.changeText(titleInput, 'Tí'));
          act(() => fireEvent(titleInput, 'blur'));

          const formErrors = UNSAFE_getAllByType(FormErrors);

          const formErrorProps = formErrors[0].props;

          expect(formErrorProps.dirty).toBe(true);
          expect(formErrorProps.error).toBe(true);
          expect(formErrorProps.errors.required).toBe(false);
          expect(formErrorProps.errors.minLength).toBe(true);
          expect(formErrorProps.errors.maxLength).toBe(false);
        });

        test('maxLength', () => {
          const { getByPlaceholderText, UNSAFE_getAllByType } = renderComp(
            props,
          );

          const titleInput = getByPlaceholderText(TITLE_INPUT_PLACEHOLDER);

          act(() =>
            fireEvent.changeText(
              titleInput,
              'Título com mais do que trinta e cinco caracteres',
            ),
          );
          act(() => fireEvent(titleInput, 'blur'));

          const formErrors = UNSAFE_getAllByType(FormErrors);

          const formErrorProps = formErrors[0].props;

          expect(formErrorProps.dirty).toBe(true);
          expect(formErrorProps.error).toBe(true);
          expect(formErrorProps.errors.required).toBe(false);
          expect(formErrorProps.errors.minLength).toBe(false);
          expect(formErrorProps.errors.maxLength).toBe(true);
        });
      });

      describe('body errors', () => {
        test('required', () => {
          const { getByPlaceholderText, UNSAFE_getAllByType } = renderComp(
            props,
          );

          const bodyInput = getByPlaceholderText(BODY_INPUT_PLACEHOLDER);

          act(() => fireEvent(bodyInput, 'blur'));

          const formErrors = UNSAFE_getAllByType(FormErrors);

          const formErrorProps = formErrors[1].props;

          expect(formErrorProps.dirty).toBe(true);
          expect(formErrorProps.error).toBe(true);
          expect(formErrorProps.errors.required).toBe(true);
          expect(formErrorProps.errors.minLength).toBe(false);
          expect(formErrorProps.errors.maxLength).toBe(false);
        });

        test('minLength', () => {
          const { getByPlaceholderText, UNSAFE_getAllByType } = renderComp(
            props,
          );

          const bodyInput = getByPlaceholderText(BODY_INPUT_PLACEHOLDER);

          act(() => fireEvent.changeText(bodyInput, 'Corpo do post'));
          act(() => fireEvent(bodyInput, 'blur'));

          const formErrors = UNSAFE_getAllByType(FormErrors);

          const formErrorProps = formErrors[1].props;

          expect(formErrorProps.dirty).toBe(true);
          expect(formErrorProps.error).toBe(true);
          expect(formErrorProps.errors.required).toBe(false);
          expect(formErrorProps.errors.minLength).toBe(true);
          expect(formErrorProps.errors.maxLength).toBe(false);
        });

        test('maxLength', () => {
          const { getByPlaceholderText, UNSAFE_getAllByType } = renderComp(
            props,
          );

          const bodyInput = getByPlaceholderText(BODY_INPUT_PLACEHOLDER);

          act(() =>
            fireEvent.changeText(
              bodyInput,
              'Corpo do post com mais de 280 caracteres, Corpo do post com mais de 280 caracteres, Corpo do post com mais de 280 caracteres, Corpo do post com mais de 280 caracteres, Corpo do post com mais de 280 caracteres, Corpo do post com mais de 280 caracteres, Corpo do post com mais de 280 caracteres, Corpo do post com mais de 280 caracteres',
            ),
          );
          act(() => fireEvent(bodyInput, 'blur'));

          const formErrors = UNSAFE_getAllByType(FormErrors);

          const formErrorProps = formErrors[1].props;

          expect(formErrorProps.dirty).toBe(true);
          expect(formErrorProps.error).toBe(true);
          expect(formErrorProps.errors.required).toBe(false);
          expect(formErrorProps.errors.minLength).toBe(false);
          expect(formErrorProps.errors.maxLength).toBe(true);
        });
      });

      describe('valid inputs', () => {
        test('valid title', () => {
          const { getByPlaceholderText, UNSAFE_getAllByType } = renderComp(
            props,
          );

          const titleInput = getByPlaceholderText(TITLE_INPUT_PLACEHOLDER);

          act(() => fireEvent.changeText(titleInput, 'Novo post'));
          act(() => fireEvent(titleInput, 'blur'));

          const formErrors = UNSAFE_getAllByType(FormErrors);

          const formErrorProps = formErrors[0].props;

          expect(formErrorProps.dirty).toBe(true);
          expect(formErrorProps.error).toBe(false);
          expect(formErrorProps.errors.required).toBe(false);
          expect(formErrorProps.errors.minLength).toBe(false);
          expect(formErrorProps.errors.maxLength).toBe(false);
        });

        test('valid body', () => {
          const { getByPlaceholderText, UNSAFE_getAllByType } = renderComp(
            props,
          );

          const bodyInput = getByPlaceholderText(BODY_INPUT_PLACEHOLDER);

          act(() =>
            fireEvent.changeText(
              bodyInput,
              'Esse é um corpo de post válido. Tem mais de 35 caracteres e menos de 280',
            ),
          );
          act(() => fireEvent(bodyInput, 'blur'));

          const formErrors = UNSAFE_getAllByType(FormErrors);

          const formErrorProps = formErrors[1].props;

          expect(formErrorProps.dirty).toBe(true);
          expect(formErrorProps.error).toBe(false);
          expect(formErrorProps.errors.required).toBe(false);
          expect(formErrorProps.errors.minLength).toBe(false);
          expect(formErrorProps.errors.maxLength).toBe(false);
        });
      });
    });
  });
});
