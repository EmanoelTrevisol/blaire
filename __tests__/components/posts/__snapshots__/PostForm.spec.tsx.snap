// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostForm.tsx matches snapshot 1`] = `
Array [
  <View
    style={
      Object {
        "marginBottom": 60,
      }
    }
  >
    <View
      style={
        Object {
          "marginVertical": 20,
          "maxHeight": 80,
        }
      }
    >
      <View
        style={
          Object {
            "marginVertical": 5,
          }
        }
      >
        <Text
          style={
            Object {
              "fontSize": 20,
            }
          }
        >
          Título
        </Text>
      </View>
      <TextInput
        allowFontScaling={true}
        autoCapitalize="none"
        editable={true}
        keyboardType="email-address"
        onBlur={[Function]}
        onChangeText={[Function]}
        placeholder="Crie seu título"
        placeholderTextColor="#898989"
        rejectResponderTermination={true}
        style={
          Object {
            "borderBottomColor": "#000",
            "borderBottomWidth": 1,
            "paddingVertical": 5,
          }
        }
        textContentType="username"
        underlineColorAndroid="transparent"
        value=""
      />
    </View>
    <View
      style={
        Object {
          "marginVertical": 20,
          "maxHeight": 80,
        }
      }
    >
      <View
        style={
          Object {
            "marginVertical": 5,
          }
        }
      >
        <Text
          style={
            Object {
              "fontSize": 20,
            }
          }
        >
          Post
        </Text>
      </View>
      <TextInput
        allowFontScaling={true}
        editable={true}
        multiline={true}
        numberOfLines={5}
        onBlur={[Function]}
        onChangeText={[Function]}
        placeholder="Mal posso esperar para ler o que você tem para contar"
        placeholderTextColor="#898989"
        rejectResponderTermination={true}
        style={
          Object {
            "borderBottomColor": "#000",
            "borderBottomWidth": 1,
            "paddingVertical": 5,
          }
        }
        underlineColorAndroid="transparent"
        value=""
      />
    </View>
  </View>,
  <View
    style={Object {}}
  >
    <View
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "backgroundColor": "#ea94a5",
          "borderRadius": 50,
          "height": 50,
          "opacity": 1,
          "padding": 16,
        }
      }
    >
      <Text
        style={
          Object {
            "fontWeight": "bold",
            "textAlign": "center",
          }
        }
      >
        Enviar
      </Text>
    </View>
  </View>,
]
`;
